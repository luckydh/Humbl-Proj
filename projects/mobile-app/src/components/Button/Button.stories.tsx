import React from "react";
import { Story, Meta } from "@storybook/react/types-6-0";

import { Button, ButtonProps } from "./Button";
import withPadding from "../../utils/withPadding";
import { withDesign } from "storybook-addon-designs";

export default {
  title: "Components/Button",
  component: Button,
  parameters: {
    design: {
      type: "figma",
      url:
        "https://www.figma.com/file/QPSFMyF0BndoFizvebnr9b/HUMBL-Pay-Mobile?node-id=539%3A307",
    },
  },
  argTypes: {
    children: { control: "text" },
    onClick: { action: "onClick" },
    variant: { control: "select" },
  },
  decorators: [withPadding, withDesign],
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  children: "Button Text",
};
export const ButtonSquare = Template.bind({});
ButtonSquare.args = {
  variant: "square",
  children: "Button Text",
};
export const ButtonSmall = Template.bind({});
ButtonSmall.args = {
  size: "small",
  children: "Button Text",
};
export const ButtonLarge = Template.bind({});
ButtonLarge.args = {
  size: "large",
  children: "Button Text",
};
export const ButtonLargeSquare = Template.bind({});
ButtonLargeSquare.args = {
  size: "large",
  variant: "square",
  children: "Button Text",
};

export const ButtonText = Template.bind({});
ButtonText.args = {
  variant: "text",
  children: "Button Text",
};
export const ButtonOutline = Template.bind({});
ButtonOutline.args = {
  variant: "outline",
  children: "Button Text",
};
export const ButtonSubmit = Template.bind({});
ButtonSubmit.args = {
  type: "submit",
  children: "Button Text",
};
export const ButtonDisabled = Template.bind({});
ButtonDisabled.args = {
  isDisabled: true,
  children: "Button Text",
};
