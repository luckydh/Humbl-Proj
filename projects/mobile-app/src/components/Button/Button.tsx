import React, { FC, CSSProperties } from "react";
import cx from "classnames";

export type ButtonVariant = "default" | "square" | "text" | "outline" | "custom";
export type ButtonSize = "default" | "large" | "small" | "xsmall";

export interface ButtonProps {
  /**
   * Disable the button
   * @default false
   * */
  isDisabled?: boolean;
  /** Respond to button click */
  onClick?: () => void;
  /**
   * Change the button styling
   * @default default
   * */
  variant?: ButtonVariant;
  children?: React.ReactNode;
  type?: "button" | "submit" | "reset";
  size?: ButtonSize;
  className?: string;
  customClass?: string;
  style?: CSSProperties;
  testId?: string;
  ariaLabel?: string;
}

const VARIANTS: Record<ButtonVariant, string> = {
  default:
    "justify-center font-medium w-full text-white py-2 px-4 bg-blue-dark rounded-md active:bg-blue-light active:shadow-button-active active:text-blue-dark transition-all",
  square:
    "rounded-lg bg-blue-dark active:bg-blue-light active:shadow-button-active border-2 border-solid border-blue-dark",
  text: "justify-center font-medium w-full text-white py-2 px-4 active:shadow-button-active transition-all",
  outline: "rounded-md font-medium w-full text-white py-2 px-4 shadow-button-outline active:bg-white-faded",
  custom: "",
};

const SIZES: Record<ButtonSize, string> = {
  default: "py-3 px-6 text-md sm:leading-none",
  large: "text-xl py-4 px-10",
  small: "py-2 px-4 text-sm",
  xsmall: "text-xs",
};

export const Button: FC<ButtonProps> = ({
  children,
  isDisabled = false,
  onClick,
  variant = "default",
  size = "default",
  type = "button",
  className,
  testId,
  ariaLabel,
  customClass,
  style,
}) => {
  const onClickHandler = () => {
    if (isDisabled) {
      return;
    }

    onClick?.();
  };

  const classes =
    customClass ??
    cx(
      "text-white select-none outline-none flex justify-center items-center",
      VARIANTS[variant],
      SIZES[size],
      isDisabled ? "pointer-events-none opacity-60" : "",
      className
    );

  return (
    <button
      type={type}
      aria-label={ariaLabel}
      data-testid={testId}
      className={classes}
      onClick={onClickHandler}
      style={variant === "default" && size === "default" ? { minWidth: 160 } : style}>
      {children}
    </button>
  );
};

export default Button;

Button.displayName = "Button";
