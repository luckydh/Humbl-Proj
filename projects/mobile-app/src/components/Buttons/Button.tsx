import { IonButton } from "@ionic/react";
import React from "react";

import "./Button.scss";

export interface ButtonProps {
  label: string;
  disabled?: boolean;
  type?: "button" | "submit" | "reset" | undefined;
  color?: string;
  fill?: any;

  /**
   * Optional click handler
   */
  onClick?: () => void;
}

const Button: React.FC<ButtonProps> = ({
  label,
  color,
  fill,
  disabled = false,
  type,
  onClick,
}: ButtonProps) => (
    <IonButton
      className="btn-customised"
      disabled={disabled}
      color={color}
      type={type}
      fill={fill}
      onClick={onClick}
    >
      {label}
    </IonButton>
  );

export default Button;
