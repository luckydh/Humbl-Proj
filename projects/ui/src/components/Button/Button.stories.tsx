import { Story, Meta } from "@storybook/react/types-6-0";
import { withDesign } from "storybook-addon-designs";
import { action } from "@storybook/addon-actions";
import { INITIAL_VIEWPORTS, MINIMAL_VIEWPORTS } from "@storybook/addon-viewport";

import { withPadding } from "utils/withPadding";

import { Button, ButtonProps } from "./Button";
import { ReactNode } from "react";

export default {
  title: "Button",
  component: Button,
  decorators: [withPadding, withDesign],
  args: {
    disabled: false,
    block: false,
    onClick: action("Click"),
  },
  parameters: {
    backgrounds: {
      default: "white",
      values: [{ name: "white", value: "#ffffff" }],
    },
    design: {
      type: "figma",
      url: "https://www.figma.com/file/QPSFMyF0BndoFizvebnr9b/HUMBL-Pay-Mobile?node-id=10332%3A100263",
    },
    viewport: {
      viewports: { ...INITIAL_VIEWPORTS, ...MINIMAL_VIEWPORTS },
      defaultViewport: "mobile1",
    },
  },
} as Meta;

type StoryProps = ButtonProps & { children: ReactNode };

const Template: Story<StoryProps> = (args) => <Button {...args} />;

const GridTemplate: Story = () => (
  <div className="uikit-grid uikit-grid-cols-2 uikit-grid-rows-1 uikit-gap-4">
    <div>
      <Button type="primary">Primary</Button>
    </div>
    <div>
      <Button type="secondary">Secondary</Button>
    </div>
  </div>
);

export const Primary = Template.bind({});
Primary.args = {
  type: "primary",
  size: "sm",
  children: "Primary Button",
};

export const Secondary = Template.bind({});
Secondary.args = {
  type: "secondary",
  size: "sm",
  children: "Secondary Button",
};

export const Tertiary = Template.bind({});
Tertiary.args = {
  type: "tertiary",
  size: "sm",
  children: "Tertiary Button",
};

export const Link = Template.bind({});
Link.args = {
  type: "link",
  size: "sm",
  children: "Link Button",
};

export const Grid = GridTemplate.bind({});
