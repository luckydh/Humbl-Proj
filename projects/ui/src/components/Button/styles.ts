import { ButtonSizes, ButtonTypes } from "./Button";

export const FULL_WIDTH = "uikit-w-full";

export const BASE_STYLES =
  "uikit-flex uikit-justify-center uikit-items-center uikit-rounded-md uikit-transition-all uikit-duration-300 uikit-border-2 uikit-outline-none uikit-relative uikit-overflow-hidden focus:uikit-outline-none focus:uikit-shadow-none disabled:uikit-opacity-50";

export const styles: Record<ButtonTypes, string> = {
  primary: "uikit-bg-blue-400 uikit-text-white uikit-border-blue-400",
  secondary: "uikit-text-blue-400 uikit-bg-blue-100 uikit-border-blue-100",
  tertiary: "uikit-bg-transparent uikit-text-blue-400 uikit-border-blue-400",
  link: "uikit-text-blue-400 uikit-border-transparent",
};

export const sizes: Record<ButtonSizes, string> = {
  lg: "uikit-px-6 uikit-py-3",
  md: "uikit-px-5 uikit-py-2.5",
  sm: "uikit-px-4 uikit-py-1.5",
};
