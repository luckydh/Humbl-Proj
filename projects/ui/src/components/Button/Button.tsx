import cx from "classnames";

import { sizes, styles, BASE_STYLES, FULL_WIDTH } from "./styles";

export type ButtonTypes = "primary" | "secondary" | "tertiary" | "link";
export type ButtonSizes = "lg" | "md" | "sm";

export interface ButtonProps {
  type?: ButtonTypes;
  size?: ButtonSizes;
  disabled?: boolean;
  block?: boolean;
  onClick?: () => void;
}

export const Button: React.FC<ButtonProps> = ({
  children,
  disabled,
  onClick,
  block = false,
  type = "primary",
  size = "sm",
}) => {
  const typeStyles = styles[type];
  const sizeStyles = sizes[size];

  return (
    <button
      onClick={onClick}
      disabled={disabled}
      className={cx(BASE_STYLES, typeStyles, sizeStyles, {
        [FULL_WIDTH]: block,
      })}
    >
      {children}
    </button>
  );
};
