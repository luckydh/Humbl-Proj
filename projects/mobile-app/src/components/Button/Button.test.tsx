import React from "react";
import renderer from "react-test-renderer";
import Button from "./Button";

it("renders default pill button correctly", () => {
  const tree = renderer.create(<Button>Default Button</Button>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders square pill button correctly", () => {
  const tree = renderer.create(<Button variant="square">Square Button</Button>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders large square button correctly", () => {
  const tree = renderer
    .create(
      <Button variant="square" size="large">
        Large Square Button
      </Button>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders large pill button correctly", () => {
  const tree = renderer.create(<Button size="large">Large Button</Button>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders text button correctly", () => {
  const tree = renderer.create(<Button variant="text">Text Button</Button>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders outline button correctly", () => {
  const tree = renderer
    .create(<Button variant="outline">Outline Button</Button>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders submit button correctly", () => {
  const tree = renderer.create(<Button type="submit">Submit Button</Button>).toJSON();
  expect(tree).toMatchSnapshot();
});
