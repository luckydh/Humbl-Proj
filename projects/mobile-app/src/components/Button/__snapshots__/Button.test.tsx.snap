// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default pill button correctly 1`] = `
<button
  className="text-white select-none outline-none flex justify-center items-center justify-center font-medium w-full text-white py-2 px-4 bg-blue-dark rounded-md active:bg-blue-light active:shadow-button-active active:text-blue-dark transition-all py-3 px-6 text-md sm:leading-none"
  onClick={[Function]}
  style={
    Object {
      "minWidth": 160,
    }
  }
  type="button"
>
  Default Button
</button>
`;

exports[`renders large pill button correctly 1`] = `
<button
  className="text-white select-none outline-none flex justify-center items-center justify-center font-medium w-full text-white py-2 px-4 bg-blue-dark rounded-md active:bg-blue-light active:shadow-button-active active:text-blue-dark transition-all text-xl py-4 px-10"
  onClick={[Function]}
  type="button"
>
  Large Button
</button>
`;

exports[`renders large square button correctly 1`] = `
<button
  className="text-white select-none outline-none flex justify-center items-center rounded-lg bg-blue-dark active:bg-blue-light active:shadow-button-active border-2 border-solid border-blue-dark text-xl py-4 px-10"
  onClick={[Function]}
  type="button"
>
  Large Square Button
</button>
`;

exports[`renders outline button correctly 1`] = `
<button
  className="text-white select-none outline-none flex justify-center items-center rounded-md font-medium w-full text-white py-2 px-4 shadow-button-outline active:bg-white-faded py-3 px-6 text-md sm:leading-none"
  onClick={[Function]}
  type="button"
>
  Outline Button
</button>
`;

exports[`renders square pill button correctly 1`] = `
<button
  className="text-white select-none outline-none flex justify-center items-center rounded-lg bg-blue-dark active:bg-blue-light active:shadow-button-active border-2 border-solid border-blue-dark py-3 px-6 text-md sm:leading-none"
  onClick={[Function]}
  type="button"
>
  Square Button
</button>
`;

exports[`renders submit button correctly 1`] = `
<button
  className="text-white select-none outline-none flex justify-center items-center justify-center font-medium w-full text-white py-2 px-4 bg-blue-dark rounded-md active:bg-blue-light active:shadow-button-active active:text-blue-dark transition-all py-3 px-6 text-md sm:leading-none"
  onClick={[Function]}
  style={
    Object {
      "minWidth": 160,
    }
  }
  type="submit"
>
  Submit Button
</button>
`;

exports[`renders text button correctly 1`] = `
<button
  className="text-white select-none outline-none flex justify-center items-center justify-center font-medium w-full text-white py-2 px-4 active:shadow-button-active transition-all py-3 px-6 text-md sm:leading-none"
  onClick={[Function]}
  type="button"
>
  Text Button
</button>
`;
